mantis-control-plane/mantis-control-plane-server/src/main/java/io/mantisrx/control/plane/resource/cluster/proto/MantisResourceClusterSpec.java

```diff
@@ -0,0 +1,153 @@
+
+    @NonNull
+    String name;
+
+    /**
+     * ID fields maps to cluster name or spinnaker app name.
+     */
+    @NonNull
+    String id;
+
+    @NonNull
+    String ownerName;
+
+    @NonNull
+    String ownerEmail;
+
+    @NonNull
+    MantisResourceClusterEnvType envType;
+
+    @Singular
+    Set<SkuTypeSpec> skuSpecs;
+
+    @Singular
+    Map<String, String> clusterMetadataFields;
+
+    @JsonCreator
+    public MantisResourceClusterSpec(
+            @JsonProperty("name") final String name,
+            @JsonProperty("id") final String id,
+            @JsonProperty("ownerName") final String ownerName,
+            @JsonProperty("ownerEmail") final String ownerEmail,
+            @JsonProperty("envType") final MantisResourceClusterEnvType envType,
+            @JsonProperty("skuSpecs") final Set<SkuTypeSpec> skuSpecs,
+            @JsonProperty("clusterMetadataFields") final Map<String, String> clusterMetadataFields) {
+        this.name = name;
+        this.id = id;
+        this.ownerName = ownerName;
+        this.ownerEmail = ownerEmail;
+        this.envType = envType;
+        this.skuSpecs = skuSpecs;
+        this.clusterMetadataFields = clusterMetadataFields;
+    }
+
+    @Builder
+    @Value
+    @EqualsAndHashCode(onlyExplicitlyIncluded = true)
+    public static class SkuTypeSpec {
+        @NonNull
+        @EqualsAndHashCode.Include
+        String skuId;
+
+        @NonNull
+        SkuCapacity capacity;
+
+        @NonNull
+        String imageId;
+
+        int cpuCoreCount;
+
+        int memorySizeInBytes;
+
+        int networkMbps;
+
+        int diskSizeInBytes;
+
+        @Singular
+        Map<String, String> skuMetadataFields;
+
+        @JsonCreator
+        public SkuTypeSpec(
+                @JsonProperty("skuId") final String skuId,
+                @JsonProperty("capacity") final SkuCapacity capacity,
+                @JsonProperty("imageId") final String imageId,
+                @JsonProperty("cpuCoreCount") final int cpuCoreCount,
+                @JsonProperty("memorySizeInBytes") final int memorySizeInBytes,
+                @JsonProperty("networkMbps") final int networkMbps,
+                @JsonProperty("diskSizeInBytes") final int diskSizeInBytes,
+                @JsonProperty("skuMetadataFields") final Map<String, String> skuMetadataFields) {
+            this.skuId = skuId;
+            this.capacity = capacity;
+            this.imageId = imageId;
+            this.cpuCoreCount = cpuCoreCount;
+            this.memorySizeInBytes = memorySizeInBytes;
+            this.networkMbps = networkMbps;
+            this.diskSizeInBytes = diskSizeInBytes;
+            this.skuMetadataFields = skuMetadataFields;
+        }
+    }
+
+    @Builder
+    @Value
+    public static class SkuCapacity {
+        @NonNull
+        String skuId;
+
+        int minSize;
+
+        int maxSize;
+
+        int desireSize;
+
+        @JsonCreator
+        public SkuCapacity(
```

Comment:
This is a workaround (when I tested the original clean version (@value + builder only) in nfmantis the shaded classes and Jackson doesn't play well with each other and other internal SDKs and it can only work when I have both annotation + upgrade shaded Jackson for all combinations).

Reply:
Let me add some notes here about this.

Reply:
Can you add some tests to make sure you codify the expected behaviors? In this case, I'm guessing you want the classes to be Jackson serializable (which they already are because of our lombok.config checked in at the root level).

Reply:
The tricky part is that these work just fine without any of the extra annotations in OSS repo but started to run into trouble imported in the internal wrapper mixing with other non-shaded Jackson. I will have more tests covering in the nfmantis side.

Reply:
For OSS I will have some extra coverage on thee too in the follow up PR.